crates/red_knot_python_semantic/src/types/property_tests.rs

```diff
@@ -123,14 +123,61 @@ impl Arbitrary for Ty {
     }
 
     fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
-        // This is incredibly naive. We can do much better here by
-        // trying various subsets of the elements in unions, tuples,
-        // and intersections. For now, we only try to shrink by
-        // reducing unions/tuples/intersections to a single element.
         match self.clone() {
-            Ty::Union(types) => Box::new(types.into_iter()),
-            Ty::Tuple(types) => Box::new(types.into_iter()),
-            Ty::Intersection { pos, neg } => Box::new(pos.into_iter().chain(neg)),
+            Ty::Union(types) => Box::new(types.shrink().filter_map(|elts| match elts.len() {
+                0 => None,
```

Comment:
It might not be completely unreasonable to try and return `Ty::Never` here?

This also makes me think: I think we currently only try shrinking for the root of the `Type` tree? Like if we have `tuple[A | B, C]`, I think we would only try to remove either `A | B` from the tuple, or try to remove `C` from the tuple, i.e. we try to shrink to `C` or to `A | B`, but we would not try to shrink the `A | B` union.

Reply:
Calling shrink on the vector of types actually does attempt to shrink the elements of the vector as well as trying to remove elements, so we get nice shrinking just leaning on the shrinking built into quickcheck (at least that was my read of the source)

Reply:
Regarding the empty union shrinking to `Ty::Never`... feels fine by me! Especially now that I fully grok that `Ty` is a smart constructor for `Type` so I don't need to worried about types being in the "wrong location", so to speak

Reply:
> Calling shrink on the vector of types actually does attempt to shrink the elements of the vector as well as trying to remove elements

:+1: 

> Regarding the empty union shrinking to `Ty::Never`... feels fine by me! Especially now that I fully grok that `Ty` is a smart constructor for `Type` so I don't need to worried about types being in the "wrong location", so to speak

If we try to shrink at all "depths", I think we can skip this. The idea was to simplify nested `Ty` unions like `A | B | (C | D)` to `A | B | Never` and then further down to `A | B`, but there are many other possible shrinking-paths that would lead to that result (e.g. via the single-element union or via a element-removal shrink on the outer union).